core/src/code.rs

```diff
@@ -99,53 +88,122 @@ impl Code {
                     }
                     _ => None,
                 })
-                .ok_or(CodeError::CheckError)?,
+                .ok_or(CodeError::MemoryEntryNotFound)?,
         );
 
+        let gas_rules = get_gas_rules(&module);
         let instrumented_module = wasm_instrument::gas_metering::inject(module, &gas_rules, "env")
             .map_err(|_| CodeError::GasInjection)?;
 
         let instrumented = wasm_instrument::parity_wasm::elements::serialize(instrumented_module)
             .map_err(|_| CodeError::Encode)?;
 
-        let code_hash = CodeId::generate(&original_code);
-
         Ok(Self {
-            original_code: Some(original_code),
             code: instrumented,
+            raw_code,
             static_pages,
             instruction_weights_version: version,
-            code_hash,
         })
     }
 
     /// Returns the original code.
-    pub fn original_code(&self) -> Option<&Vec<u8>> {
-        self.original_code.as_ref()
+    pub fn raw_code(&self) -> &[u8] {
+        &self.raw_code
     }
 
     /// Returns reference to the instrumented binary code.
     pub fn code(&self) -> &[u8] {
         &self.code
     }
 
-    /// Returns code hash.
-    pub fn code_hash(&self) -> CodeId {
-        self.code_hash
+    /// Returns instruction weights version.
+    pub fn instruction_weights_version(&self) -> u32 {
+        self.instruction_weights_version
     }
 
-    /// Set code hash.
-    pub fn set_code_hash(&mut self, code_hash: CodeId) {
-        self.code_hash = code_hash;
+    /// Returns initial memory size from memory import.
+    pub fn static_pages(&self) -> WasmPageNumber {
+        self.static_pages
+    }
+
+    /// Consumes this instance and returns the raw binary code.
+    pub fn into_raw_code(self) -> Vec<u8> {
+        self.raw_code
+    }
+}
+
+/// The newtype contains the Code instance and the corresponding id (hash).
+#[derive(Clone, Debug)]
+pub struct CodeAndId(Code, CodeId);
+
+impl CodeAndId {
+    /// Calculates the id (hash) of the raw binary code and creates new instance.
+    pub fn new(code: Code) -> Self {
+        let code_id = CodeId::generate(code.raw_code());
+        Self(code, code_id)
+    }
+
+    /// Creates the instance from the precalculated hash without checks.
+    pub fn from_parts_unchecked(code: Code, code_id: CodeId) -> Self {
+        assert_eq!(code_id, CodeId::generate(code.raw_code()));
+        Self(code, code_id)
+    }
+
+    /// Returns corresponding id (hash) for the code.
+    pub fn code_id(&self) -> CodeId {
+        self.1
+    }
+
+    /// Returns reference to Code.
+    pub fn code(&self) -> &Code {
+        &self.0
+    }
+
+    /// Decomposes this instance.
+    pub fn into_parts(self) -> (Code, CodeId) {
+        (self.0, self.1)
+    }
+}
+
+/// The newtype contains the instrumented code and the corresponding id (hash).
+#[derive(Clone, Debug, Decode, Encode)]
+pub struct InstrumentedCodeAndId(Vec<u8>, WasmPageNumber, u32, CodeId);
```

Comment:
Isn't a structure better here instead of a tuple?

Reply:
Yep, maybe better. However the struct is intentionally designed not to expose its members: they should be accessed with the corresponding getters. So the struct with named field would have no advantage over the tuple.

Reply:
You'll potentially get problems after adding some element to the middle of the tuple.